sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,591 @@
+                return Err(Error::File {
+                    verb: "create directory",
+                    path: temp_dir,
+                    err,
+                });
+            }
+        }
+        let (file, temp_path) = tokio::task::spawn_blocking(move || {
+            NamedUtf8TempFile::new_in(&temp_dir).map_err(|err| Error::File {
+                verb: "create temporary file in",
+                path: temp_dir,
+                err,
+            })
+        })
+        .await??
+        .into_parts();
+        let file =
+            NamedUtf8TempFile::from_parts(File::from_std(file), temp_path);
+
+        let (mut file, hasher) = stream
+            .try_fold(
+                (file, Sha256::new()),
+                |(mut file, mut hasher), chunk| async move {
+                    hasher.update(&chunk);
+                    match file.as_file_mut().write_all(chunk.as_ref()).await {
+                        Ok(()) => Ok((file, hasher)),
+                        Err(err) => Err(Error::File {
+                            verb: "write to",
+                            path: file.path().to_owned(),
+                            err,
+                        }),
+                    }
+                },
+            )
+            .await?;
+        file.as_file_mut().sync_data().await.map_err(|err| Error::File {
+            verb: "sync",
+            path: final_path.clone(),
+            err,
+        })?;
+
+        let digest = hasher.finalize();
+        if digest.as_slice() != sha256.as_ref() {
+            return Err(Error::HashMismatch {
+                expected: sha256,
+                actual: ArtifactHash(digest.into()),
+            });
+        }
+
+        let moved_final_path = final_path.clone();
+        tokio::task::spawn_blocking(move || {
+            file.persist(&moved_final_path).map_err(|err| Error::File {
+                verb: "rename temporary file to",
+                path: moved_final_path,
+                err: err.error,
+            })
+        })
+        .await??;
+        info!(
+            &self.log,
+            "Wrote artifact";
+            "sha256" => &sha256.to_string(),
+            "path" => final_path.as_str(),
+        );
+        Ok(())
+    }
+
+    /// PUT operation (served by Sled Agent API) which takes a [`StreamingBody`]
+    pub(crate) async fn put_body(
+        &self,
+        sha256: ArtifactHash,
+        body: StreamingBody,
+    ) -> Result<(), Error> {
+        self.put(sha256, body.into_stream().map_err(Error::Body)).await
+    }
+
+    /// POST operation (served by Sled Agent API)
+    pub(crate) async fn copy_from_depot(
+        &self,
+        sha256: ArtifactHash,
+        address: SocketAddrV6,
+    ) -> Result<(), Error> {
+        static CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
+            reqwest::ClientBuilder::new()
+                .connect_timeout(Duration::from_secs(15))
+                .read_timeout(Duration::from_secs(15))
```

Comment:
> The fact that the copy-from-depot operation is synchronous and these timeouts being so short make me think the implementation assumes the whole process of fetching the artifact will be fairly quick.

[`read_timeout`](https://docs.rs/reqwest/0.12.8/reqwest/struct.ClientBuilder.html#method.read_timeout) here is actually the timeout between individual reads: "The timeout applies to each read operation, and resets after a successful read. This is more appropriate for detecting stalled connections when the size isn’t known beforehand."

But yes I think you're right to question this call being synchronous. This is making me realize that Nexus does not care if this copy ever completes, because I suspect the flow I will write for the Nexus background task is:

1. Ask all the sled agents what artifacts they have
2. For each sled agent and for each artifact, if that sled agent is missing that artifact, tell them to get it from a sled agent that claims to have it
3. For each sled agent, if they have artifacts we don't have in our list, tell that sled agent to delete it
4. Probably also don't tell 29 sled agents to download 100 new artifacts all at once when a new repo is uploaded (maybe shuffle the list of things to do and only perform the first N operations)

If Nexus gets an error from that sled, it will log it and then do nothing with it, and then try again next time the background task kicks off, only after it has asked each sled what artifacts it has. Nexus will eventually find out what the result of the operation was without having to specifically poll the result of this response.

One case where maybe Nexus cares about the error is if we tell it "hey, the other sled agent gave us garbage data", where its response would be to tell that sled agent to delete its file and then eventually fetch it again from somewhere else. (Is that likely to happen?) But to guard against this, the remote sled agent can check that itself and delete its own file if it finds out the hash is wrong as it serves the file.

In the other potential case, where Nexus is waiting for the sled to fetch an artifact so that it can tell the sled to apply that update, it could instead keep calling the "copy artifact" endpoint if we implement the change that you suggest next:

>  This solves a few problems: if there's an interruption, or if Nexus just wants to give up, it can just try the request again. That will never create extra load and it will still correctly wait for the fetch to complete. But also: if two Nexus instances try to sync the same artifact to the same sled at the same time (which is not hard to imagine, since both are doing this constantly in the background), we won't do 2x the work.
> 
> But as I write this, it seems like you might also want to have these operations check for the presence of the object locally first. Otherwise, we could still wind up copying the same file over and over again in these cases.

I think:

- We should detect if we already have the artifact in question and return some 2XX response. If Nexus really wants sled agent to re-fetch the artifact it can ask it to delete it first.
- We should be able to detect if we're already trying to write the artifact in question and return a `409 Conflict`. (Should this count direct PUTs of the artifact? I suppose so.)

Reply:
So this is mostly done now:

- The temporary file is not random but instead `tmp/{sha256}`. We use `OpenOptions::create_new` to ensure that we're the only task trying to write to that object; otherwise we return 409 Conflict.
- In `copy_from_depot`, after conflicts are checked and upstream request are sent, we move the actual copy into a task and return 200 OK.

We aren't checking to see if the file already exists. I wrote above:

> If Nexus really wants sled agent to re-fetch the artifact it can ask it to delete it first.

But given the change to writing artifacts to both M.2s, there can be a situation where an I/O error prevents writing an artifact to one volume but it succeeds on the other. We want Nexus to be aware of this situation and be able to remediate it, so I changed the `list` method to return a count of each artifact present; the expectation is that if the count is wrong, Nexus will ask that sled agent to refetch the artifact so it can be present on both disks.